### PR TITLE
Enable child row selection in TableNext

### PR DESCRIPTION
**Pull request by MariaJose**
## What?
1. Enable child row selection  (isChildrenRowsSelectable)
2. Get rows by and identifier (getRowId)
3. Add and update props. (initialSelectedParentRows, expandedRowSelector)

Note: Please review by commit.

## Why?
1. Let the user decide to make all the children selectable in a table or not.
2. Let the user decide how to retrieve the selected rows and expanded rows by an identifier (the  default is gonna by indexes)
3. Based on the new changes of how to retrieve selected rows, it was necessary to add initialSelectedParentRows prop to identify the initial selected parent rows across pages, since when using pagination not all the items are passed to the component, it is just passes the current items by page. 
Also, update expandedRowSelector prop name to getChildren since it better describes the purpose of it.

New props:
`isChildrenRowsSelectable`:
  - type: boolean
  - description: enables child row selection. If the prop is not used, by default child rows are not enabled
  - example: 
  
  ```
    selectable={
       isChildrenRowsSelectable: true
    }
  ```
  
`getRowId`:
  - type: (item: T, parentRowId: number, childRowId: number) => string
  - description: a way to identify a row by a unique id, sku, etc. Note: The default is by indexes.
  - example: If all the items contains a unique identifier (example, id, sku, etc) , the user would need to provide a callback function in to the prop:
   
   ```ts
   const children = [{name: 'child 1', id: 'id-child1'}, {name: 'child 2',  id: 'id-child2'}]
   const items = [{name: 'item 1', id: 'id-item1', children}]
   getRowById: {(item) => item.id}
   
   // Result after selecting all items
   {
      id-item1: true,
      id-child1: true,
      id-child2: true
   }

   // if getRowId is not used, the default selected items would be:
   
   {
       0: true
       0.0: true
       0.1: true
    }
   ```

`initialSelectedParentRows`
  -type: string[]
  -description: This helps the component to internally count the all the total selected parent rows selected cross pages.
  -example: 
  
  ```
   const [selectedItems, onSelectionChange] = useState<Record<string, true>>({0: true,  20: true, 20.0:  true});
   selectable: {
       selectedItems,
       initialSelectedParentRows: ['0', '20']
   }
  ```
  
 Updated props:

 `expandedRowSelector`
  - what changed: Only the name,  NOTE: the usage of this prop didn't change, it's still the same
  - new name: getChildren.
  - description: this prop helps table next to retrieve children from an item by passing a callback function.
  - example:
  ```ts
  expandable={{
        getChildren: ({ children }) => {
          if (children) {
            return children;
          }

          return undefined;
        },
      }}
  ```
  
Code example:

```tsx
import { TableNext } from '@bigcommerce/big-design/src/components/TableNext';
import React, { useEffect, useState } from 'react';

interface Item {
  name: string;
  id: string;
  age: number;
  children?: Item[];
}

const itemsWithChilds = new Array(25).fill(undefined).map<Item>((_, index) => ({
  name: `Hello ${index}`,
  id: `id-${index}`,
  age: index + 1,
  children: [
    { name: `Child ${index}`, id: `id-${index}-0`, age: index + 1 },
    { name: `Child ${index}`, id: `id-${index}-1`, age: index + 2 },
  ],
}));

const itemWithNoChild = new Array(1).fill(undefined).map<Item>((_, index) => ({
  name: `Hello item with no child`,
  id: `id-item-with-no-child`,
  age: index + 1,
}));

const items = [...itemWithNoChild, ...itemsWithChilds];

const DevPage = () => {
  const [currentPage, onPageChange] = useState(1);
  const [itemsPerPage, onItemsPerPageChange] = useState(10);
  const [selectedItems, onSelectionChange] = useState<Record<string, true>>({
    'id-1': true,
    'id-1-0': true,
    'id-2': true,
    'id-13': true,
    'id-13-0': true,
  });
  const [pagedItems, setPagedItems] = useState<typeof items>([]);
  const [expandedRows, onExpandedChange] = useState({});

  useEffect(() => {
    setPagedItems(
      items.filter(
        (_item, index) =>
          index >= (currentPage - 1) * itemsPerPage && index < currentPage * itemsPerPage,
      ),
    );
  }, [currentPage, itemsPerPage]);

  return (
    <TableNext
      columns={[
        { hash: 'name', header: 'Name', render: ({ name }) => name },
        { hash: 'age', header: 'Age', render: ({ age }) => age },
      ]}
      expandable={{
        expandedRows,
        onExpandedChange,
        getChildren: ({ children }) => {
          if (children) {
            return children;
          }

          return undefined;
        },
      }}
      itemName="Worlds"
      items={pagedItems}
      keyField="sku"
      pagination={{
        currentPage,
        onPageChange,
        itemsPerPage,
        totalItems: items.length,
        onItemsPerPageChange,
        itemsPerPageOptions: [10, 20, 30],
      }}
      selectable={{
        onSelectionChange,
        selectedItems,
        isChildrenRowsSelectable: true,
        initialSelectedParentRows: ['id-1', 'id-13'],
      }}
      getRowId={(item) => {
        return item.id;
      }}
    />
  );
};

export default DevPage;
```
## Screenshots/Screen Recordings
Children are selectable. Selecting children by a unique identifier (id):

https://user-images.githubusercontent.com/39739180/211925719-b1ea96e2-a805-4dda-be98-c10a504201f3.mp4

https://user-images.githubusercontent.com/39739180/211927411-d0bab745-4702-4fd9-a0f6-24da7751074e.mp4

Parent are just selectable. Selecting parents by default (indexes):

https://user-images.githubusercontent.com/39739180/211925950-0d51d79d-b883-4384-8c5d-9d9ae5874df1.mp4

## Testing/Proof
Tests pass

### CODE REVIEW

**Review by jorgemoya — APPROVED** [review]
LGTM but it is definitely hard to follow! Lets see if we can some extra 👀 , otherwise we can merge and keep testing.

**Review by eugene-polev — APPROVED** [review]

**Comment by MariaJose** [comment]
> Looks mostly good to me, but the thing I'm noticing is that some of the new internal props (eg `getChildren`, things that deal with row ids, etc) seem to be passed along between components quite often
> 
> Maybe it makes sense to put some of these things into a "TableNext" context? That way we would avoid passing them along on all these subcomponents, as they would be readily available in hooks, utils, etc. The tradeoff would be of course that sharing things globally like that comes with its own complications (we'd need to be kinda smart about what to put there and what not to put there, as it could get out of hand). Just a thought, but curious what you think

discussed offline, for the moment we are gonna keep it as it is since having a context would save us a level

**Review by golcinho — APPROVED** [review]

**Review by rtalvarez — APPROVED** [review]
